runtime/js-compute-runtime/builtins/cache-simple.cpp

```diff
@@ -0,0 +1,440 @@
+    JS_FN("json", bodyAll<RequestOrResponse::BodyReadResult::JSON>, 0, JSPROP_ENUMERATE),
+    JS_FN("text", bodyAll<RequestOrResponse::BodyReadResult::Text>, 0, JSPROP_ENUMERATE),
+    JS_FS_END,
+};
+
+const JSPropertySpec SimpleCacheEntry::properties[] = {
+    JS_PSG("body", body_get, JSPROP_ENUMERATE),
+    JS_PSG("bodyUsed", bodyUsed_get, JSPROP_ENUMERATE),
+    JS_STRING_SYM_PS(toStringTag, "SimpleCacheEntry", JSPROP_READONLY),
+    JS_PS_END,
+};
+
+bool SimpleCacheEntry::constructor(JSContext *cx, unsigned argc, JS::Value *vp) {
+  JS_ReportErrorUTF8(cx, "SimpleCacheEntry can't be instantiated directly");
+  return false;
+}
+
+JSObject *SimpleCacheEntry::create(JSContext *cx, fastly_body_handle_t body_handle) {
+  JS::RootedObject SimpleCacheEntry(cx, JS_NewObjectWithGivenProto(cx, &class_, proto_obj));
+  if (!SimpleCacheEntry)
+    return nullptr;
+
+  JS::SetReservedSlot(SimpleCacheEntry, static_cast<uint32_t>(Slots::Body),
+                      JS::Int32Value(body_handle));
+  JS::SetReservedSlot(SimpleCacheEntry, static_cast<uint32_t>(Slots::BodyStream), JS::NullValue());
+  JS::SetReservedSlot(SimpleCacheEntry, static_cast<uint32_t>(Slots::HasBody),
+                      JS::BooleanValue(true));
+  JS::SetReservedSlot(SimpleCacheEntry, static_cast<uint32_t>(Slots::BodyUsed), JS::FalseValue());
+
+  return SimpleCacheEntry;
+}
+
+bool SimpleCacheEntry::init_class(JSContext *cx, JS::HandleObject global) {
+  return init_class_impl(cx, global);
+}
+
+namespace {
+// We currently support five types of body inputs:
+// - byte sequence
+// - buffer source
+// - USV strings
+// - URLSearchParams
+// - ReadableStream (Currently only supports Host-backed ReadableStreams)
+// After the other other options are checked explicitly, all other inputs are
+// encoded to a UTF8 string to be treated as a USV string.
+// TODO: Add support for Blob and FormData when we have implemented those classes.
+JS::Result<std::tuple<JS::UniqueChars, size_t>> convertBodyInit(JSContext *cx,
+                                                                JS::HandleValue bodyInit) {
+  JS::RootedObject bodyObj(cx, bodyInit.isObject() ? &bodyInit.toObject() : nullptr);
+  mozilla::Maybe<JS::AutoCheckCannotGC> maybeNoGC;
+  JS::UniqueChars buf;
+  size_t length;
+
+  if (bodyObj && JS_IsArrayBufferViewObject(bodyObj)) {
+    // `maybeNoGC` needs to be populated for the lifetime of `buf` because
+    // short typed arrays have inline data which can move on GC, so assert
+    // that no GC happens. (Which it doesn't, because we're not allocating
+    // before `buf` goes out of scope.)
+    maybeNoGC.emplace(cx);
+    JS::AutoCheckCannotGC &noGC = maybeNoGC.ref();
+    bool is_shared;
+    length = JS_GetArrayBufferViewByteLength(bodyObj);
+    buf = JS::UniqueChars(
+        reinterpret_cast<char *>(JS_GetArrayBufferViewData(bodyObj, &is_shared, noGC)));
+  } else if (bodyObj && JS::IsArrayBufferObject(bodyObj)) {
+    bool is_shared;
```

Comment:
Our runtime does not include support for SharedArrayBuffers, meaning this would/should always be false, I will add a MOZ_ASSERT to ensure that is the case 👍